### ISSUE TITLE
Suggestion: useHead

### ISSUE THREAD

**Issue by Ephem** [JoviDeCroock/hoofd#15]
Hey, nice work on this!

I'm planning out a possible refactor of how we handle stuff in `<head>` and a possibly nice addition to this library struck me. Several apps I've worked with in the past has represented all this data in an object, building that object up in a Redux selector or by some other method, then passing that `head`-object to a specialised component that maps it to Helmet.

Maybe there is an opportunity here to get rid of boilerplate in userland in these cases by providing a `useHead`-hook that is a combination of all the others and takes an object, or array, as an argument?

This was also inspired by [Remix](https://remix.run/), where returning a `{ meta: {} }` from preload-functions will be supported out of the box.

**Comment by JoviDeCroock** [comment]
Hey @Ephem 	

Welcome and thanks!

I do like the idea behind this but it feels pretty hard to keep performant. I guess most of the time this array would be fresh on every render meaning we’d have to traverse this and check it with the DOM. In our plain useMeta, useLink,... we can take advantage of the flat structure of the argument to prevent this check from being needed.

There are ways to possibly circumvent this but I fear these might contradict the “rules of hooks” for instance iterating over said array and just using the hooks but if the array becomes larger/smaller this will induce errors in React.

Lastly I think this might also become a significant increase in size if we’d do the first approach (call dispatcher for meta/title/titleTemplate) and manual for links (since these can be duped).

What are your thoughts on this?

**Comment by Ephem** [comment]
Those are valid points!

I guess it's one of those cumbersome "you should memoize this in userland" kind of scenarios, which wont work because people don't read the docs or like to memoize things. :P Then again, if the component where hooked head is used does not re-render often, it's probably not a problem to start with, but it does push that complexity to userland either way.

Without being familiar with the implementation I guess one way to at least avoid DOM access would be to keep track of the last applied value for this particular tag?

While I do feel this probably has the same complexity regardless of being implemented in a library or directly in apps (like we do now), I'm definitely emphatic to the payload increase. This library is very lean right now which would be a shame to increase if this is functionality that few people use and I'm guessing this would be hard to build in a way that it minifies away if unused. :)

**Comment by JoviDeCroock** [comment]
I've been giving this more thought and this might be the ideal use case for a [`Babel Macro`](https://github.com/kentcdodds/babel-plugin-macros) this could allow us to export a hook called `useHead` which spawns a function that keeps the length of the previous execution and compares it to the current execution. When it sees we're adding a new meta tag it will throw and suggest an empty placeholder for it.

```
const macro = (arguments) => {
  arguments.forEach(() => {
    if (meta) useMeta()
  });
}
```

**Comment by Ephem** [comment]
To avoid breaking the rule of hooks? I think the problem with that is that metadata is often generated from dynamic data not known at build time, so you can't always predict which placeholders to add.

**Comment by JoviDeCroock** [comment]
Since this is a development tool it only warns the user, it does not attempt to predict what the new values will be.

It only transforms `useHead` into a series of the native hooks.

**Comment by Ephem** [comment]
But it does this at build time right?

We have a component that stays mounted between page transitions which could theoretically use this hook. We fetch the metadata for each page from an API at runtime so arguments are not known ahead of time, and they change in length between different pages. Some pages have og-ratings and some pages do not for example, but in reality a whole bunch of things is removed and added.

**Comment by JoviDeCroock** [comment]
Hmm, the goal for me with this library was to allow for granularity as opposed to, how I perceived, massive blocks.

```jsx
<helm>
  <meta />
  ...
</helm>
```

Where here it would be a `useMeta/...` on a subcomponent for granular control.

**Comment by Ephem** [comment]
I'm not quite sure I understood your last comment fully, but I totally get that you want the API lean and focused on granularity as opposed to what might very well be a pretty niche usecase, that can also be pretty easily built on top. :smile:

I wasn't sure if it was a good fit myself and just thought I'd spin the idea by you, thanks for taking the time to discuss!

**Comment by JoviDeCroock** [comment]
@Ephem no need to close this, I do agree on it being a nice feature, the rules of hooks however make this a very tall feat because, every render should have the same amount of hooks.

**Comment by Ephem** [comment]
Oh sorry, I misunderstood, I thought you were saying those hindrances made it hard to fit this feature into this library. I'll reopen. 😄 

I think it would be very hard to have this as a grouped hook for that reason. I've thought some more about this, and maybe leaving memoization up to the user isn't such a bad approach after all? In the bigger picture, I think memoization needs to become the standard when passing objects to custom hooks or every single custom hook out there will have this problem. Doing deep compare by default because users _might_ forget to memoize will probably hurt more in the long run.

If you think that is too much responsibility to put on the user, I think either keeping track of the last values that was queued to the dispatcher in the `useHead` hook and avoiding queuing them if they are the same OR keeping track of the last values that were added/updated to the DOM inside the dispatcher itself and avoiding adding/updating them if they are still the same is probably the best way to go. If you add that logic to the `useHead` hook itself and that is properly tree-shakeable, that might mean the payload doesn't have to go up for users not using it?

**Comment by JoviDeCroock** [comment]
Implemented in `hoofd@1.0.0`
